### PR TITLE
refactor: simplify dispatching logic

### PR DESCRIPTION
**Pull request by osdnk**

### CODE REVIEW

File: packages/core/src/useNavigationHelpers.tsx [hunk 1]
```diff
@@ -82,4 +94,27 @@ export default function useNavigationHelpers<
     setState,
     onAction,
   ]);
+
+  const { addNavigationListener } = React.useContext(NavigationFocusedContext);
```
Comment (satya164) [inline packages/core/src/useNavigationHelpers.tsx]: I think it's better to move this code to a separate hook. `useNavigationHelpers` is to create navigation helper object, so it's not the right place for this. Maybe `useFocusedListeners`

File: packages/core/src/NavigationFocusedContext.tsx [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+export type FocusedNavigationCallback = (
+  navigation: NavigationHelpers<ParamListBase>
+) => any;
+
+export type FocusedNavigationListener = (
+  callback: FocusedNavigationCallback
+) => { handled: boolean; result: any };
+
+const MISSING_CONTEXT_ERROR =
+  "We couldn't find a navigation focused context. Have you wrapped your app with 'NavigationContainer'?";
```
Comment (satya164) [inline packages/core/src/NavigationFocusedContext.tsx]: Maybe we should't throw this error? A "navigation focused context" is an internal thing, so the error message could be confusing for the user

File: packages/core/src/NavigationFocusedContext.tsx [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+/**
+ * Context which holds the required helpers needed to access focused navigator.
+ */
+
+const NavigationFocusedContext = React.createContext<{
```
Comment (satya164) [inline packages/core/src/NavigationFocusedContext.tsx]: I've removed the need for a `removeXListener`, so maybe we can use the same builder context for this now?

File: packages/core/src/NavigationFocusedContext.tsx [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    throw new Error(MISSING_CONTEXT_ERROR);
+  },
+});
+
+export default NavigationFocusedContext;
+
+/**
+ * Hook which lets child navigators add action listeners.
+ */
+export function useFocusedListeners() {
```
Comment (satya164) [inline packages/core/src/NavigationFocusedContext.tsx]: Let's move this hook to separate file like other hooks. We could move the listener logic you added in `useNavigationHelpers` into this hook because they are related logic.

File: packages/core/src/NavigationFocusedContext.tsx [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  "We couldn't find a navigation focused context. Have you wrapped your app with 'NavigationContainer'?";
+
+/**
+ * Context which holds the required helpers needed to access focused navigator.
+ */
+
+const NavigationFocusedContext = React.createContext<{
+  addNavigationListener: (listener: FocusedNavigationListener) => void;
+}>({
+  get addNavigationListener(): any {
```
Comment (satya164) [inline packages/core/src/NavigationFocusedContext.tsx]: Not sure about `addNavigationListener` since we're not listening to navigation. Maybe something like `addFocusedListener` to match the naming of the context and the hook?

File: packages/core/src/useNavigationBuilder.tsx [hunk 6]
```diff
@@ -18,6 +18,7 @@ import {
   Router,
   RouterFactory,
 } from './types';
+import { useFocusedListeners } from './NavigationFocusedContext';
```
Comment (satya164) [inline packages/core/src/useNavigationBuilder.tsx]: nit: types at the end :D

File: packages/core/src/useNavigationHelpers.tsx [hunk 7]
```diff
@@ -11,6 +11,10 @@ import {
   ParamListBase,
   Router,
 } from './types';
+import NavigationFocusedContext, {
```
Comment (satya164) [inline packages/core/src/useNavigationHelpers.tsx]: nit: types at the end

File: packages/core/src/useNavigationHelpers.tsx [hunk 8]
```diff
@@ -21,6 +25,7 @@ type Options<State extends NavigationState, Action extends NavigationAction> = {
   setState: (state: State) => void;
   emitter: NavigationEventEmitter;
   router: Router<State, Action>;
+  navigationListeners: FocusedNavigationListener[];
```
Comment (satya164) [inline packages/core/src/useNavigationHelpers.tsx]: Perhaps `focusedListeners` following naming of the hook and the context?

File: packages/core/src/useNavigationHelpers.tsx [hunk 9]
```diff
@@ -82,4 +94,27 @@ export default function useNavigationHelpers<
     setState,
     onAction,
   ]);
+
+  const { addNavigationListener } = React.useContext(NavigationFocusedContext);
+  const listener = React.useRef((callback: FocusedNavigationCallback) => {
+    if (navigation.isFocused()) {
```
Comment (satya164) [inline packages/core/src/useNavigationHelpers.tsx]: I was wondering if the `target` approach we following in `onAction` is better than checking `navigation.isFocused` as `isFocused` will check focused state of each parent. but lets leave for separate PR

File: packages/core/src/NavigationContainer.tsx [hunk 10]
```diff
@@ -92,10 +93,10 @@ const Container = React.forwardRef(function NavigationContainer(
     getPartialState(initialState)
   );
 
+  const { listeners, addNavigationListener } = useFocusedListeners();
+
   const dispatch = (action: NavigationAction) =>
-    context.performTransaction(() => {
-      actionListeners[0](action, undefined, null);
-    });
+    listeners[0](navigation => navigation.dispatch(action));
```
Comment (satya164) [inline packages/core/src/NavigationContainer.tsx:99]: Still think there's no reason not to loop over the listeners.

File: packages/core/src/NavigationFocusedContext.tsx [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+  callback: FocusedNavigationCallback
+) => { handled: boolean; result: any };
+
+const MISSING_CONTEXT_ERROR =
+  "We couldn't find a navigation focused context. Have you wrapped your app with 'NavigationContainer'?";
+
+/**
+ * Context which holds the required helpers needed to access focused navigator.
+ */
+
```
Comment (satya164) [inline packages/core/src/NavigationFocusedContext.tsx]: I think there shouldn't be new line here or JSDoc hint won't work

File: packages/core/src/useNavigationHelpers.tsx [hunk 12]
```diff
@@ -82,4 +94,27 @@ export default function useNavigationHelpers<
     setState,
     onAction,
   ]);
+
+  const { addNavigationListener } = React.useContext(NavigationFocusedContext);
+  const listener = React.useRef((callback: FocusedNavigationCallback) => {
```
Comment (satya164) [inline packages/core/src/useNavigationHelpers.tsx]: Why `useRef` here? Maybe you wanted to use `useCallback`? Using ref is incorrect since the values will never update.

File: packages/core/src/useNavigationHelpers.tsx [hunk 13]
```diff
@@ -82,4 +94,27 @@ export default function useNavigationHelpers<
...
+      return { handled: true, result: callback(navigation) };
+    } else {
+      return { handled: false, result: null };
+    }
+  });
+
+  React.useEffect(() => {
+    const { current: currentListener } = listener;
+    return addNavigationListener(currentListener);
+  });
```
Comment (satya164) [inline packages/core/src/useNavigationHelpers.tsx]: Need to add a dependency array to avoid the effect running every render.

**Review by satya164 — CHANGES_REQUESTED** [review]

File: packages/core/src/NavigationContainer.tsx (hunk 10 above)
Comment (satya164) [reply packages/core/src/NavigationContainer.tsx:99]: Should wrap this in curly braces so the return value of `dispatch` isn't a boolean

File: packages/core/src/__tests__/useFocusEffect.test.tsx [hunk 14]
```diff
@@ -45,17 +45,23 @@ it('runs focus effect on focus change', () => {
...
+  act(() => {
+    navigation.current.navigate('second');
+  });
 
   expect(focusEffect).toBeCalledTimes(1);
   expect(focusEffectCleanup).not.toBeCalled();
 
-  act(() => navigation.current.navigate('third'));
+  act(() => {
+    navigation.current.navigate('third');
```
Comment (satya164) [inline packages/core/src/__tests__/useFocusEffect.test.tsx]: I think we should revert this change because it caught an issue where we were leaking internal return value as the return value of `dispatch`

**Review by satya164 — CHANGES_REQUESTED** [review]

File: packages/core/src/NavigationFocusedContext.tsx (hunk 2 above)
Comment (osdnk) [reply packages/core/src/NavigationFocusedContext.tsx]: I think that it's not important. I just didn't want to make keys optional.

File: packages/core/src/NavigationFocusedContext.tsx (hunk 3 above)
Comment (osdnk) [reply packages/core/src/NavigationFocusedContext.tsx]: can you clarify?

Comment (osdnk) [reply packages/core/src/NavigationFocusedContext.tsx]: Oh you mean to construct useEffect directly here. It makes sense, but I don't have opinion if it's obviously better

File: packages/core/src/useNavigationHelpers.tsx (hunk 9 above)
Comment (osdnk) [reply packages/core/src/useNavigationHelpers.tsx]: no strong opinion

File: packages/core/src/NavigationContainer.tsx (hunk 10 above)
Comment (osdnk) [reply packages/core/src/NavigationContainer.tsx:99]: I think since we're absolutely sure that here cannot be more than one listener it's ok

File: packages/core/src/NavigationFocusedContext.tsx (hunk 3 above)
Comment (osdnk) [reply packages/core/src/NavigationFocusedContext.tsx]: got it! thanks!

File: packages/core/src/NavigationFocusedContext.tsx (hunk 2 above)
Comment (osdnk) [reply packages/core/src/NavigationFocusedContext.tsx]: maybe I can change default to: `() => {}`

Comment (osdnk) [reply packages/core/src/NavigationFocusedContext.tsx]: Or whatever, it can be optional. Fixed!
